### DISCUSSION TITLE
Using flow control characteristics

### DISCUSSION THREAD

**Discussion by lechndo** [Jakeler/ble-serial#43] [Ideas]
Hallo Mr. Jakeler,

I really want to thank you for this nice piece of software.

I just wanted to ask if you are planning to integrate support for the usage of "flow control characteristics"?
This would be great to prevent data loss.

The different approaches I found:
- u-Blox: https://www.u-blox.com/sites/default/files/u-connectXpress-LowEnergySerialPortService_ProtocolSpec_UBX-16011192.pdf
- Telit: https://www.telit.com/wp-content/uploads/2017/10/Terminal_IO_Profile_Client_Implementation_Guide_r7.pdf
- Laird: http://cdn.lairdtech.com/home/brandworld/files/Application%20Note%20-%20Using%20VSP%20with%20smartBASIC.pdf

All of the use some characteristics to give the remote side a certain number of "_credits_", defining the amount of data being allowed so send.

I will need a reliable connection and data transfer for my next microcontroller project at work and also for the related PC testers.
So if I don't find a different - already working - solution I maybe will create a fork and try to implement it on my own... if I get some time for it during the project.

Kind regards

**Comment by Jakeler** [comment]
Hi @lechndo,
using additional characteristics for flow control is new to me, I have not planned to support it so far. Though it looks interesting, thanks for the suggestion! I will look into it if I have more spare time, but can't promise anything. Feel free to open a draft PR if you start to implement it, then I am happy to help and merge this in.

Do you already have devices that support it? 
I would also need at least one to develop and test the connection, probably programming an ESP32 would be the easiest option.

What are your bandwidth requirements? 
Maybe you have seen my performance tests here: https://blog.ja-ke.tech/2021/04/22/ble-serial-2.html#performance
It is already very reliable if you make sure that the application stays below 10 kbit/s in general. At least with the standard BLE/UART modules, which basically have no buffer and lose received data if the uart output can't keep up, this could be surely improved with custom firmware.

**Reply by lechndo** [reply]
Thx for the quick reply.

Currently we use a "Telit BlueMod+SR", a dual mode module. I'm not sure if it is still available but there are similar modules like the "BlueMod+S42" which also support the "Telit Terminal I/O" service.
For the upcoming project we plan to use the "uBlox NINA-B41".
And we also have an eval kit for the NINA-B41... [EVK-NINA-B416](https://www.mouser.at/ProductDetail/u-blox/EVK-NINA-B416?qs=%2Fha2pyFadujAo1ZhnqISSpBnYBkDPrB0JrdhpDKST1pm2s5oMOdBrA%3D%3D)

Both, and also all other modules that use "credits" for flow control (at least those I had a closer look on), also use/support hardware flow control on the external UART interface. Otherwise it wouldn't make much sense... ;-)

Our bandwidth requirements are relatively moderate. The default of 115k of our modules is quite enough but we have to send several kBs of data want to avoid loosing any data. And we also don't want to have to care about the remote sides buffer sizes etc.
So a working flow control mechanism will be important.

It will also take some time until I get to the point where I will have t write some testers (which will be in python), but I really hope that I then will have some time to implement this feature... If it's not already done... :-) 
If you start with it before I do, I would be glad to help or answer any question you might have.

**Comment by blauret** [comment]
The [Dialog DSPS ](https://www.dialog-semiconductor.com/products/dialog-serial-port-service-dsps)(my company) is using flow control tool. 

We're probably interested in add the support for the flow control.
